gradle-plugin/src/main/java/com/squareup/anvil/plugin/AnvilPlugin.kt

```diff
@@ -88,233 +142,123 @@ open class AnvilPlugin : KotlinCompilerPluginSupportPlugin {
-          )
-        }
-      }
   }
 
-  private fun disableIncrementalKotlinCompilation(
-    project: Project,
-    projectType: ProjectType,
-    extension: AnvilExtension
-  ) {
+  private fun disableIncrementalKotlinCompilation(variant: Variant) {
     // Use this signal to share state between DisableIncrementalCompilationTask and the Kotlin
     // compile task. If the plugin classpath changed, then DisableIncrementalCompilationTask sets
     // the signal to false.
     @Suppress("UnstableApiUsage")
-    val incrementalSignal = IncrementalSignal.registerIfAbsent(project)
+    val incrementalSignal = IncrementalSignal.registerIfAbsent(variant.project)
 
-    val taskSuffix = when (projectType) {
-      JVM, ANDROID -> "Kotlin"
-      MULTIPLATFORM_JVM -> "KotlinJvm"
-      MULTIPLATFORM_ANDROID -> "KotlinAndroid"
-    }
+    disableIncrementalCompilationWhenTheCompilerClasspathChanges(
+      variant, incrementalSignal, variant.compileTaskProvider
+    )
 
-    if (extension.generateDaggerFactoriesOnly.get() ||
-      extension.disableComponentMerging.get()
-    ) {
-      // We don't need to disable the incremental compilation for the stub generating task, when we
-      // only generate Dagger factories or contributing modules. That's only needed for merging
-      // Dagger modules.
-      if (projectType.isAndroid) {
-        project.androidVariantsConfigure { variant ->
-          val compileTaskName = "kaptGenerateStubs${variant.name.capitalize(US)}$taskSuffix"
-          disableIncrementalCompilationAction(project, incrementalSignal, compileTaskName)
-        }
-      } else {
-        disableIncrementalCompilationAction(
-          project,
-          incrementalSignal,
-          "kaptGenerateStubs$taskSuffix"
-        )
-        disableIncrementalCompilationAction(
-          project,
-          incrementalSignal,
-          "kaptGenerateStubsTest$taskSuffix"
-        )
-      }
-    } else {
-      project.tasks
-        .withType(KaptGenerateStubsTask::class.java)
-        .configureEach { stubsTask ->
-          // Disable incremental compilation for the stub generating task. Trigger the compiler
-          // plugin if any dependencies in the compile classpath have changed. This will make sure
-          // that we pick up any change from a dependency when merging all the classes. Without
-          // this workaround we could make changes in any library, but these changes wouldn't be
-          // contributed to the Dagger graph, because incremental compilation tricked us.
-          stubsTask.doFirstCompat {
-            stubsTask.incremental = false
-            stubsTask.logger.info(
-              "Anvil: Incremental compilation enabled: ${stubsTask.incremental} (stub)"
+    variant.project.pluginManager.withPlugin("org.jetbrains.kotlin.kapt") {
+      variant.project
+        .namedLazy<KaptGenerateStubsTask>(
+          "kaptGenerateStubs${variant.taskSuffix}"
```

Comment:
I have to do this, because this
```kotlin
.namedLazy<KaptGenerateStubsTask>("kaptGenerateStubs${variant.taskSuffix}") { stubsTaskProvider ->
```
is too long.